test/distributed/pipelining/test_backward.py

```diff
@@ -53,6 +57,127 @@ def test_stage_backward(self):
                 print(f"Gradient test failed for {name}: {p.grad} vs {ref_p.grad}")
                 raise
 
+    def test_stage_backward_input(self):
+        # MLP as a stage module
+        mod = MLPModule(d_hid)
+        x = torch.randn(batch_size, d_hid)
+        # As in a pipeline stage, the inputs to this stage requires gradients
+        x.requires_grad_(True)
+        target = torch.randn(batch_size, d_hid)
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        # Make a copy
+        ref_mod = copy.deepcopy(mod)
+        ref_x = x.detach().requires_grad_(x.requires_grad)
+        ref_target = target.detach()
+
+        # Forward, then backward of loss with respect to inputs
+        out = mod(x)
+        loss = loss_fn(out, target)
+        dinputs, param_groups = stage_backward_input(
+            stage_outputs=(loss,),
+            output_grads=None,
+            stage_inputs=[x],
+            weights=mod.parameters(),
+        )
+
+        # Run reference
+        ref_out = ref_mod(ref_x)
+        ref_loss = loss_fn(ref_out, ref_target)
+        ref_loss.backward()
+
+        torch.testing.assert_close(x.grad, ref_x.grad)
```

Comment:
nit: would be worth also asserting that mod.parameters grads are None?

Reply:
good idea, added